pkg/integrations/prometheus/utils/operatorutils.go

```diff
@@ -0,0 +1,243 @@
+package utils
+
+import (
+	"fmt"
+	"math"
+	"regexp"
+	"sort"
+
+	"github.com/prometheus/prometheus/promql/parser"
+	"github.com/siglens/siglens/pkg/segment/structs"
+	"github.com/siglens/siglens/pkg/segment/utils"
+	segutils "github.com/siglens/siglens/pkg/segment/utils"
+	log "github.com/sirupsen/logrus"
+)
+
+func GetLogicalAndArithmeticOperation(op parser.ItemType) segutils.LogicalAndArithmeticOperator {
+	switch op {
+	case parser.ADD:
+		return segutils.LetAdd
+	case parser.SUB:
+		return segutils.LetSubtract
+	case parser.MUL:
+		return segutils.LetMultiply
+	case parser.DIV:
+		return segutils.LetDivide
+	case parser.MOD:
+		return segutils.LetModulo
+	case parser.POW:
+		return segutils.LetPower
+	case parser.GTR:
+		return segutils.LetGreaterThan
+	case parser.GTE:
+		return segutils.LetGreaterThanOrEqualTo
+	case parser.LSS:
+		return segutils.LetLessThan
+	case parser.LTE:
+		return segutils.LetLessThanOrEqualTo
+	case parser.EQLC:
+		return segutils.LetEquals
+	case parser.NEQ:
+		return segutils.LetNotEquals
+	case parser.LAND:
+		return segutils.LetAnd
+	case parser.LOR:
+		return segutils.LetOr
+	case parser.LUNLESS:
+		return segutils.LetUnless
+	default:
+		log.Errorf("getArithmeticOperation: unexpected op: %v", op)
+		return 0
+	}
+}
+
+func IsLogicalOperator(op segutils.LogicalAndArithmeticOperator) bool {
+	switch op {
+	case segutils.LetAnd:
+		return true
+	case segutils.LetOr:
+		return true
+	case segutils.LetUnless:
+		return true
+	default:
+		return false
+	}
+}
+
+func SetFinalResult(queryOp structs.QueryArithmetic, finalResult map[string]map[uint32]float64, groupID string, timestamp uint32, valueLHS float64, valueRHS float64, swapped bool) {
+
+	if swapped {
+		tmp := valueLHS
+		valueLHS = valueRHS
+		valueRHS = tmp
+	}
+
+	if queryOp.ReturnBool {
+		finalResult[groupID][timestamp] = 0
+	}
+
+	switch queryOp.Operation {
+	case utils.LetAdd:
+		finalResult[groupID][timestamp] = valueLHS + valueRHS
+	case utils.LetDivide:
+		if valueRHS == 0 {
+			return
+		}
+		finalResult[groupID][timestamp] = valueLHS / valueRHS
+	case utils.LetMultiply:
+		finalResult[groupID][timestamp] = valueLHS * valueRHS
+	case utils.LetSubtract:
+		val := valueLHS - valueRHS
+		finalResult[groupID][timestamp] = val
+	case utils.LetModulo:
+		finalResult[groupID][timestamp] = math.Mod(valueLHS, valueRHS)
+	case utils.LetPower:
+		finalResult[groupID][timestamp] = math.Pow(valueLHS, valueRHS)
+	case utils.LetGreaterThan:
+		isGtr := valueLHS > valueRHS
+		if isGtr {
+			if queryOp.ReturnBool {
+				finalResult[groupID][timestamp] = 1
+				return
+			}
+			if swapped {
+				finalResult[groupID][timestamp] = valueRHS
+			} else {
+				finalResult[groupID][timestamp] = valueLHS
+			}
+		}
```

Comment:
You're still doing some swapping logic inside `case`s. Since you've moved the swapping logic above the `switch`, you need to remove it from all of the `case`s. Please go through this whole switch statement again and remove all swapping logic.

Also, why are the PR tests passing with this bug? You should probably add more unit tests that will test this issue of incorrectly swapping twice instead of once.

Reply:
I think we still need the swapped logic for some operators. For example, for comparison operators, if the equation evaluates to true, we should preserve the value of the vector `on the left`. However, since we moved the swap logic out of the switch case, we want to ensure that the preserved value is from `the left vector`.